main.cpp

```diff
@@ -261,247 +367,365 @@ struct CoatRack
+
+void ChannelStrip::acceptAudioInput(bool isConnected, bool isMono)
+{
+    acceptMicInput(false);
+    isConnected = true;
+    isMono = false;
+}
+
+//============================================================
+
 struct Equalizer
 {
-    // 1) number of bands
     int numBands = 3;
-    // 2) max range of frequency
     double maxFreqRange = 20000.0;
-    // 3) color of knobs
     std::string knobColor = "Blue";
-    // 4) type of filter
     std::string filterType = "lowShelf";
-    // 5) smoothness of control
     double knobSmoothness = 2.6;
 
-    // 1) boost the bass frequency
     double boostBass(float level, float freqRange);
-    // 2) cut the high frequency
     double cutHiFreq(float amountToCut, float freqRange);
-    // 3) adjust the width of the frequency band
-    double adjustWidth(float level);
+    double adjustWidth(double level);
 };
-/*
-Thing 7) Preamp
-5 properties:
-    1) quality of effect
-    2) number of controls
-    3) amount of gain increase
-    4) line sensitivity
-    5) clipping indicator
-3 things it can do:
-    1) boost signal amplitude 
-    2) trim signal amplitude
-    3) distort overall signal
- */
+
+double Equalizer::boostBass(float level, float freqRange)
+{
+    level = 24.0f;
+    freqRange = 94.0f;
+
+    return 0;
+}
+
+double Equalizer::cutHiFreq(float amountToCut, float freqRange)
+{
+    if (amountToCut > freqRange)
+    {
+        freqRange = 1200.0f;
+    }
+
+    return static_cast<double>(freqRange);
+}
+
+double Equalizer::adjustWidth(double level)
+{
+    if (level > 12.0)
+    {
+        level = 12.0;
+    } 
+
+    return level;
+}
+
+//============================================================
+
 struct Preamp
 {
-    // 1) quality of effect
     float quality = 96.0f;
-    // 2) number of controls
     int numControls = 1;
-    // 3) amount of gain increase
     float gainIncrease = 24.0f;
-    // 4) line sensitivity
     double lineSensitivity = 10.0;
-    // 5) clipping indicator
     bool clippingIndicator = false;
 
-    // 1) boost signal amplitude 
-    int  boostAmp(float level);
-    // 2) trim signal amplitude
-    int trimAmp(float level);
-    // 3) distort overall signal
-    int distortSignal(float level, bool allowClip = false);
+    float boostAmp(float level);
+    float trimAmp(float level);
+    float distortSignal(float level, bool allowClip = false);
 };
-/*
-Thing 8) Master Controls
-5 properties:
-    1) volume control level
-    2) panning control setting
-    3) level meter indicator
-    4) range of master volume in db
-    5) headphone level
-3 things it can do:
-    1) adjust the master volume
-    2) adjust the headphone volume
-    3) indicate overall output level
- */
+
+float Preamp::boostAmp(float level)
+{
+    level *= gainIncrease;
+
+    return level;
+}
+
+float Preamp::trimAmp(float level)
+{
+    level = gainIncrease - 5.0f;
+
+    return level;
+}
+
+float Preamp::distortSignal(float level, bool allowClip)
+{
+    if (allowClip == false)
```

Comment:
You can also write `if( !allowClip )`.  It's your call as to which one you find easier to understand.
- [ ] revise if you want.